Added `Ember.warn` helpers

slindberg opened the pull request:
Fixes #157

This PR adds the `expectWarning`, `expectNoWarning`, and `ignoreWarning` helpers whose behavior is identical to the similarly named deprecation helpers.

Along the way I implemented pending deprecation helper tests, added the ability to pass a callback to `expectNoDeprecation`, and refactored method call tracking into a util.

mixonic commented on addon/test-helper/deprecation.js:
```diff
@@ -1,53 +1,50 @@
   reset: function(){
-    this.expecteds = null;
-    this.actuals = null;
+    this.tracker = new MethodCallTracker(this.env.Ember, 'deprecate');
```
It is a little funky that `this.tracker` is used here, but `invokeWithNewContext` could be called with a completely different tracker than this one.

Can everything use the tracker on `this` for the sake of not having two floating around when one is all that is necessary?

slindberg replied:
Given the way that the helpers behave when passed a callback vs. not, I thought this turned out to be a decent abstraction, although I see why it can seem weird. Really, it just fell out of extracting the tracking logic into a util, and resulted in the removal of the actual/expected swapping logic that was pretty difficult to follow IMO.

I think the alternative to creating new instances of the tracker would be to give them something like a `createNewContext` method so that the util could internally manage the temporary context shift. I have a feeling you were thinking something specific?

mixonic replied:
I have nothing specific in mind beyond trying to keep only one tracker around for the sake on sanity.

IMO could be as simple as having `invokeWithNewContext` call a `buildTracker` method that returns one present as `this.tracker` if already built and creates a new one (and sets it) if there is none present. Then `reset` could just tear it down.

In general, I encourage learning toward patterns that are lazy. create a call tracker only if needed, and use reset to tear it down and set the stage for the next if you need to.

Vague thoughts, I know.

slindberg replied:
I'll do a little refactoring to only instantiate a tracker when necessary, but that doesn't address the issue of having to deal with multiple contexts:

```js
expectDeprecation(/foo/);

// tests that could issue deprecation 'foo'

expectDeprecation(function() {
  // test that issues deprecation 'bar'
}, /bar/);

// tests that could also issue deprecation 'foo'
```

AFAICT, `DeprecationAssert` has to deal with the fact that the function given to the second call is not expected to issue the 'foo' deprecation. Essentially there's a single global context for `expectDeprecation` and N number of sandboxed contexts for each call that is passed a function. This is why the `invokeWithNewContext` method creates a new tracker instance and then promptly drops it on the floor.

Although perhaps I've misunderstood this completely o_O

mixonic commented:
This is pretty fantastic @slindberg! Have you tried it against the Ember test suite?

slindberg commented:
@mixonic I have not, I will attempt to do so tomorrow morning.

slindberg commented:
@mixonic I refactored the helpers to lazily instantiate call trackers, but new instances are still created when a helper is invoked with a function after being invoked without one.

I ran it against [current ember master](https://github.com/emberjs/ember.js/tree/68ddfa2d02f61138c4a68f3876b7c9a91864cf1c) and everything passes.

slindberg commented:
@mixonic @mmun, Should I try to rebase with master?

mixonic commented:
@slindberg arrrg sorry for not getting to this. Yes a rebase would be very welcome, I will happily merge.

mmun commented:
@slindberg Oh no! I'm sorry I didn't realize this PR was pending. The rebase might be a bit painful.

slindberg commented:
I rebased and refactored common logic into a `DebugAssert` base class, which currently uses es6 classes. If I [whitelist `es6.classes` in the babel config](https://github.com/emberjs/emberjs-build/blob/64402929a425e1c506344a1516a743a7d36bd2cd/lib/utils/transpile-es6.js#L43-L54), all tests pass in canary.

@rwjblue wasn't opposed to adding classes to the ember build, but if anyone has concerns, I can remove usages of `class`.

slindberg commented:
@mmun I didn't want to bloat this PR any more than necessary, so I opened [a follow-up issue](https://github.com/emberjs/ember-dev/issues/163) to track the extraction of test framework specific code.

mmun commented:
Looks great to me.

rwjblue commented:
`es6.classes` added in https://github.com/emberjs/emberjs-build/pull/125 and Ember updated in https://github.com/emberjs/ember.js/pull/12301.

rwjblue commented:
Can you submit a PR to Ember updating using your fork/branch in `bower.json` so we can confirm all tests pass?  Then once green we can merge + release, then update the PR to the new merge commit SHA...

slindberg commented:
@rwjblue https://github.com/emberjs/ember.js/pull/12311

Looks like CI is failing on the `old-jquery-and-extend-prototypes` and `sauce` suites, investigating...

slindberg commented:
@rwjblue the previous issue was just a bad rebase, [but now it's failing on IE9](https://travis-ci.org/emberjs/ember.js/jobs/79737522) :neutral_face: :gun:

It looks like there's zero visibility into what failed in sauce. Do you have any additional access to logs that could help me track down the failure? I don't have a convenient way to test in IE9 unfortunately.

mmun commented:
@slindberg It was just a random failure. It's passing now.

slindberg commented:
Awesome. I noticed I forgot a couple of quotes in the assertion messages and rebased. This should be good to merge now.
